app/controllers/works_controller.rb

```diff
@@ -37,50 +40,78 @@ def show
+      redirect_to work_path(@work.id)
+      flash[:result_text] = "Only the person that added the work can edit or delete it"
+    end
   end
 
+  #only the user who created the work can update
   def update
-    @work.update_attributes(media_params)
-    if @work.save
-      flash[:status] = :success
-      flash[:result_text] = "Successfully updated #{@media_category.singularize} #{@work.id}"
-      redirect_to work_path(@work)
+    # puts "BEFORE LOGIC****"
+    # ap params
+    # puts "@work.user: #{@work.user}"
+    # puts "params[:id]: #{params[:id]}"
+    # puts "BEFORE LOGIC****"
+    unless @work.can_edit_or_delete_work?(session[:user_id])
+      # puts "unless****************"
+      # puts @work.can_edit_or_delete_work?(params[:user_id])
+      redirect_to work_path(@work.id)
+      flash[:result_text] = "Only the person that added the work can edit or delete it"
+      # puts "unless****************"
     else
-      flash.now[:status] = :failure
-      flash.now[:result_text] = "Could not update #{@media_category.singularize}"
-      flash.now[:messages] = @work.errors.messages
-      render :edit, status: :not_found
+      @work.update_attributes(media_params)
+      if @work.save
+        puts "SAVED!"
+        flash[:status] = :success
+        flash[:result_text] = "Successfully updated #{@media_category.singularize} #{@work.id}"
+        redirect_to work_path(@work.id)
+      else
+        puts "@work not saved****************"
+        ap @work
+        puts "@work not saved****************"
+        flash.now[:status] = :failure
+        flash.now[:result_text] = "Could not update #{@media_category.singularize}"
+        flash.now[:messages] = @work.errors.messages
+        render :edit, status: :not_found
+      end
     end
   end
 
   def destroy
-    @work.destroy
-    flash[:status] = :success
-    flash[:result_text] = "Successfully destroyed #{@media_category.singularize} #{@work.id}"
-    redirect_to root_path
+    if @work.can_edit_or_delete_work?(session[:user_id])
+      @work.destroy
+      flash[:status] = :success
+      flash[:result_text] = "Successfully destroyed #{@media_category.singularize} #{@work.title}"
+      redirect_to root_path
+    else
+      flash.now[:status] = :failure
+      flash.now[:result_text] = "Could not delete the work #{@work.title}"
+      redirect_to work_path(@work.id)
+    end
   end
 
   def upvote
-    # Most of these varied paths end in failure
-    # Something tragically beautiful about the whole thing
-    # For status codes, see
-    # http://stackoverflow.com/questions/3825990/http-response-code-for-post-when-resource-already-exists
-    flash[:status] = :failure
-    if @login_user
-      vote = Vote.new(user: @login_user, work: @work)
-      if vote.save
-        flash[:status] = :success
-        flash[:result_text] = "Successfully upvoted!"
-        status = :found
+    if @work.can_edit_or_delete_work?(session[:user_id])
```

Comment:
I was wrong earlier in that if you use `redirect` you can't return a status code other than 3XX, (redirection).  

You can instead render something and give a response code, or just redirect with a flash notice.